src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php

```diff
@@ -0,0 +1,249 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\Framework\Test\Api\ApiDefinition\Generator;
+
+use League\Flysystem\FileNotFoundException;
+use League\Flysystem\Filesystem;
+use Shopware\Core\Framework\Test\TestCaseBase\FilesystemBehaviour;
+use Shopware\Core\Framework\Test\TestCaseBase\KernelTestBehaviour;
+
+/**
+ * Incomplete adapter to make Flysystem usable via stream wrapper.
+ *
+ * It allows using the test Flysystem instance with file system related code expecting string paths.
+ * This class only implements the methods for reading files and directories, writes have to be
+ * done using the Flysystem instance directly.
+ */
+class FlysystemReadOnlyStreamWrapper
+{
+    use KernelTestBehaviour;
+    use FilesystemBehaviour;
+
+    const MODES = [
+        'file' => 0100644, // rw-r--r--
+        'dir' => 0040755,  // rwxr-xr-x
+    ];
+
+    /**
+     * @var string
+     */
+    private $filePath;
+
+    private $fileOffset = 0;
+
+    private $dirListing = [];
+
+    private function getFilesystemInstance(): Filesystem
+    {
+        return $this->getFilesystem('shopware.filesystem.private');
+    }
+
+    private function parseWrapperFilename(string $path): string
+    {
+        $parts = parse_url(str_replace(':///', '://', $path));
+
+        return ($parts['host'] ?? '') . ($parts['path'] ?? '');
+    }
+
+    private function getMetadata(string $filename): array
+    {
+        try {
+            return $this->getFilesystemInstance()->getMetadata($filename);
+        } catch (FileNotFoundException $exception) {
+            return [];
+        }
+    }
+
+    private function getSize(): int
+    {
+        return $this->getMetadata($this->filePath)['size'];
+    }
+
+    /**
+     * @throws FileNotFoundException
+     */
+    private function getFileContents(): string
+    {
+        return $this->getFilesystemInstance()->read($this->filePath);
+    }
+
+    public function stream_open(string $path, string $mode, int $options, &$opened_path): bool
+    {
+        $filePath = $this->parseWrapperFilename($path);
+
+        if (! $this->isFile($filePath)) {
+            return false;
+        }
+
+        $this->filePath   = $filePath;
+        $this->fileOffset = 0;
+
+        return true;
+    }
+
+    /**
+     * @throws FileNotFoundException
```

Comment:
we only annotate exceptions if they are thrown in this method

Reply:
PHPStorm doesn't like that, but done :)